File: implementations/android/client/app/src/main/java/com/theshuai/specus/android/PeerPathMtu.java
package com.theshuai.specus.android;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.function.LongSupplier;

/** Authenticated SPM2 packetization-layer MTU messages and per-path discovery state. */
final class PeerPathMtu {
    static final int MIN_INNER_MTU = 576;
    static final int MAX_INNER_MTU = 9_000;
    static final int MAX_ATTEMPTS = 3;
    static final long PROBE_TIMEOUT_MILLIS = 750L;
    static final long CACHE_TTL_MILLIS = 10L * 60_000L;

    private static final byte[] MAGIC = "SPMTU2".getBytes(StandardCharsets.US_ASCII);
    private static final int TYPE_PROBE = 1;
    private static final int TYPE_ACK = 2;
    private static final int HEADER_BYTES = MAGIC.length + 1 + Long.BYTES + Short.BYTES;

    private PeerPathMtu() {
    }

    static byte[] probe(long nonce, int innerMtu) {
        validate(nonce, innerMtu);
        byte[] payload = new byte[innerMtu];
        writeHeader(payload, TYPE_PROBE, nonce, innerMtu);
        return payload;
    }

    static byte[] ack(long nonce, int innerMtu) {
        validate(nonce, innerMtu);
        byte[] payload = new byte[HEADER_BYTES];
        writeHeader(payload, TYPE_ACK, nonce, innerMtu);
        return payload;
    }

    static boolean looksLike(byte[] payload) {
        if (payload == null || payload.length < MAGIC.length) {
            return false;
        }
        for (int index = 0; index < MAGIC.length; index++) {
            if (payload[index] != MAGIC[index]) {
                return false;
            }
        }
        return true;
    }

    static Message decode(byte[] payload) {
        if (!looksLike(payload) || payload.length < HEADER_BYTES) {
            return null;
        }
        ByteBuffer input = ByteBuffer.wrap(payload, MAGIC.length, payload.length - MAGIC.length);
        int type = input.get() & 0xFF;
        long nonce = input.getLong();
        int innerMtu = input.getShort() & 0xFFFF;
        if (nonce <= 0 || innerMtu < MIN_INNER_MTU || innerMtu > MAX_INNER_MTU) {
            return null;
        }
        if (type == TYPE_PROBE && payload.length == innerMtu) {
            return new Message(true, nonce, innerMtu);
        }
        return type == TYPE_ACK && payload.length == HEADER_BYTES
                ? new Message(false, nonce, innerMtu)
                : null;
    }

    private static void writeHeader(byte[] output, int type, long nonce, int innerMtu) {
        ByteBuffer.wrap(output)
                .put(MAGIC)
                .put((byte) type)
                .putLong(nonce)
                .putShort((short) innerMtu);
    }

    private static void validate(long nonce, int innerMtu) {
        if (nonce <= 0 || innerMtu < MIN_INNER_MTU || innerMtu > MAX_INNER_MTU) {
            throw new IllegalArgumentException("invalid path MTU message");
        }
    }

    static final class Message {
        final boolean probe;
        final long nonce;
        final int innerMtu;

        Message(boolean probe, long nonce, int innerMtu) {
            this.probe = probe;
            this.nonce = nonce;
            this.innerMtu = innerMtu;
        }
    }

    static final class Probe {
        final long nonce;
        final int innerMtu;

        Probe(long nonce, int innerMtu) {
            this.nonce = nonce;
            this.innerMtu = innerMtu;
        }
    }

    static final class Transition {
        final Probe probe;
        final Integer completedMtu;

        Transition(Probe probe, Integer completedMtu) {
            this.probe = probe;
            this.completedMtu = completedMtu;
        }

        static Transition idle() {
            return new Transition(null, null);
        }
    }

    static final class Discovery {
        private String pathKey = "";
        private int ceiling = MIN_INNER_MTU;
        private int lower = MIN_INNER_MTU;
        private int upper = MIN_INNER_MTU;
        private int effective = MIN_INNER_MTU;
        private int pendingSize;
        private long pendingNonce;
        private int attempts;
        private boolean sawFailure;
        private long validUntilMillis;

        synchronized Transition activate(String nextPathKey,
                                         int configuredMtu,
                                         Integer cachedMtu,
                                         long cachedUntilMillis,
                                         long nowMillis,
                                         LongSupplier nonceSupplier) {
            int normalized = normalize(configuredMtu);
            boolean samePath = nextPathKey != null && nextPathKey.equals(pathKey);
            if (samePath && (pendingSize > 0 || validUntilMillis > nowMillis)) {
                return Transition.idle();
            }
            pathKey = nextPathKey == null ? "" : nextPathKey;
            ceiling = normalized;
            lower = MIN_INNER_MTU;
            upper = normalized;
            pendingSize = 0;
            pendingNonce = 0L;
            attempts = 0;
            sawFailure = false;
            if (cachedMtu != null && cachedUntilMillis > nowMillis) {
                effective = Math.min(normalized, normalize(cachedMtu));
                lower = effective;
                upper = effective;
                validUntilMillis = cachedUntilMillis;
                return Transition.idle();
            }
            effective = normalized;
            validUntilMillis = 0L;
            return issue(normalized, nonceSupplier);
        }

        synchronized Transition acknowledge(long nonce,
                                            int innerMtu,
                                            long nowMillis,
                                            LongSupplier nonceSupplier) {
            if (pendingSize == 0 || pendingNonce != nonce || pendingSize != innerMtu) {
                return Transition.idle();
            }
            lower = Math.max(lower, innerMtu);
            effective = sawFailure ? lower : ceiling;
            pendingSize = 0;
            pendingNonce = 0L;
            attempts = 0;
            if (lower >= upper) {
                return complete(nowMillis);
            }
            return issue(lower + ((upper - lower + 1) / 2), nonceSupplier);
        }

        synchronized Transition timeout(long nonce, long nowMillis, LongSupplier nonceSupplier) {
            if (pendingSize == 0 || pendingNonce != nonce) {
                return Transition.idle();
            }
            if (attempts < MAX_ATTEMPTS) {
                attempts++;
                return new Transition(new Probe(pendingNonce, pendingSize), null);
            }
            sawFailure = true;
            upper = Math.max(MIN_INNER_MTU, pendingSize - 1);
            effective = Math.min(effective, upper);
            pendingSize = 0;
            pendingNonce = 0L;
            attempts = 0;
            if (upper <= lower) {
                effective = lower;
                return complete(nowMillis);
            }
            return issue(lower + ((upper - lower + 1) / 2), nonceSupplier);
        }

        synchronized int effectiveMtu(int configuredMtu) {
            return Math.min(normalize(configuredMtu), Math.max(MIN_INNER_MTU, effective));
        }

        synchronized String pathKey() {
            return pathKey;
        }

        private Transition issue(int size, LongSupplier nonceSupplier) {
            long nonce;
            do {
                nonce = nonceSupplier.getAsLong() & Long.MAX_VALUE;
            } while (nonce == 0L);
            pendingSize = Math.max(MIN_INNER_MTU, Math.min(ceiling, size));
            pendingNonce = nonce;
            attempts = 1;
            return new Transition(new Probe(nonce, pendingSize), null);
        }

        private Transition complete(long nowMillis) {
            effective = Math.max(MIN_INNER_MTU, Math.min(ceiling, effective));
            validUntilMillis = nowMillis + CACHE_TTL_MILLIS;
            return new Transition(null, effective);
        }

        private static int normalize(int mtu) {
            return Math.max(MIN_INNER_MTU, Math.min(MAX_INNER_MTU, mtu));
        }
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerPathMtu.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Specus.Client.PeerMesh;

internal static class PeerPathMtu
{
    public const int MinInnerMtu = 576;
    public const int MaxInnerMtu = 9000;
    public const int MaxAttempts = 3;
    public static readonly TimeSpan ProbeTimeout = TimeSpan.FromMilliseconds(750);
    public static readonly TimeSpan CacheTtl = TimeSpan.FromMinutes(10);

    private static readonly byte[] Magic = Encoding.ASCII.GetBytes("SPMTU2");
    private const byte ProbeType = 1;
    private const byte AckType = 2;
    private const int HeaderBytes = 6 + 1 + 8 + 2;

    internal sealed record Message(bool IsProbe, ulong Nonce, int InnerMtu);
    internal sealed record Probe(ulong Nonce, int InnerMtu);
    internal sealed record Transition(Probe? NextProbe = null, int? CompletedMtu = null);
    internal sealed record CacheEntry(int InnerMtu, DateTimeOffset ValidUntil);

    public static byte[] EncodeProbe(ulong nonce, int innerMtu)
    {
        Validate(nonce, innerMtu);
        var payload = new byte[innerMtu];
        WriteHeader(payload, ProbeType, nonce, innerMtu);
        return payload;
    }

    public static byte[] EncodeAck(ulong nonce, int innerMtu)
    {
        Validate(nonce, innerMtu);
        var payload = new byte[HeaderBytes];
        WriteHeader(payload, AckType, nonce, innerMtu);
        return payload;
    }

    public static bool LooksLike(ReadOnlySpan<byte> payload) =>
        payload.Length >= Magic.Length && payload[..Magic.Length].SequenceEqual(Magic);

    public static Message? Decode(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < HeaderBytes || !LooksLike(payload))
        {
            return null;
        }
        var type = payload[6];
        var nonce = BinaryPrimitives.ReadUInt64BigEndian(payload[7..15]);
        var innerMtu = BinaryPrimitives.ReadUInt16BigEndian(payload[15..17]);
        if (nonce == 0 || innerMtu < MinInnerMtu || innerMtu > MaxInnerMtu)
        {
            return null;
        }
        if (type == ProbeType && payload.Length == innerMtu)
        {
            return new Message(true, nonce, innerMtu);
        }
        return type == AckType && payload.Length == HeaderBytes
            ? new Message(false, nonce, innerMtu)
            : null;
    }

    private static void WriteHeader(Span<byte> output, byte type, ulong nonce, int innerMtu)
    {
        Magic.CopyTo(output);
        output[6] = type;
        BinaryPrimitives.WriteUInt64BigEndian(output[7..15], nonce);
        BinaryPrimitives.WriteUInt16BigEndian(output[15..17], checked((ushort)innerMtu));
    }

    private static void Validate(ulong nonce, int innerMtu)
    {
        if (nonce == 0 || innerMtu < MinInnerMtu || innerMtu > MaxInnerMtu)
        {
            throw new ArgumentOutOfRangeException(nameof(innerMtu), "invalid path MTU message");
        }
    }

    internal sealed class Discovery
    {
        private readonly object _sync = new();
        private string _pathKey = "";
        private int _ceiling = MinInnerMtu;
        private int _lower = MinInnerMtu;
        private int _upper = MinInnerMtu;
        private int _effective = MinInnerMtu;
        private int _pendingSize;
        private ulong _pendingNonce;
        private int _attempts;
        private bool _sawFailure;
        private DateTimeOffset _validUntil;

        public Transition Activate(string pathKey, int configuredMtu, CacheEntry? cached, DateTimeOffset now)
        {
            lock (_sync)
            {
                configuredMtu = Normalize(configuredMtu);
                if (_pathKey == pathKey && (_pendingSize > 0 || now < _validUntil))
                {
                    return new Transition();
                }
                _pathKey = pathKey;
                _ceiling = configuredMtu;
                _lower = MinInnerMtu;
                _upper = configuredMtu;
                _pendingSize = 0;
                _pendingNonce = 0;
                _attempts = 0;
                _sawFailure = false;
                if (cached is not null && now < cached.ValidUntil)
                {
                    _effective = Math.Min(configuredMtu, Normalize(cached.InnerMtu));
                    _lower = _effective;
                    _upper = _effective;
                    _validUntil = cached.ValidUntil;
                    return new Transition();
                }
                _effective = configuredMtu;
                _validUntil = default;
                return Issue(configuredMtu);
            }
        }

        public Transition Acknowledge(ulong nonce, int innerMtu, DateTimeOffset now)
        {
            lock (_sync)
            {
                if (_pendingSize == 0 || _pendingNonce != nonce || _pendingSize != innerMtu)
                {
                    return new Transition();
                }
                _lower = Math.Max(_lower, innerMtu);
                _effective = _sawFailure ? _lower : _ceiling;
                _pendingSize = 0;
                _pendingNonce = 0;
                _attempts = 0;
                if (_lower >= _upper)
                {
                    return Complete(now);
                }
                return Issue(_lower + ((_upper - _lower + 1) / 2));
            }
        }

        public Transition Timeout(ulong nonce, DateTimeOffset now)
        {
            lock (_sync)
            {
                if (_pendingSize == 0 || _pendingNonce != nonce)
                {
                    return new Transition();
                }
                if (_attempts < MaxAttempts)
                {
                    _attempts++;
                    return new Transition(new Probe(_pendingNonce, _pendingSize));
                }
                _sawFailure = true;
                _upper = Math.Max(MinInnerMtu, _pendingSize - 1);
                _effective = Math.Min(_effective, _upper);
                _pendingSize = 0;
                _pendingNonce = 0;
                _attempts = 0;
                if (_upper <= _lower)
                {
                    _effective = _lower;
                    return Complete(now);
                }
                return Issue(_lower + ((_upper - _lower + 1) / 2));
            }
        }

        public int EffectiveMtu(int configuredMtu)
        {
            lock (_sync)
            {
                return Math.Min(Normalize(configuredMtu), Math.Max(MinInnerMtu, _effective));
            }
        }

        public string PathKey
        {
            get
            {
                lock (_sync)
                {
                    return _pathKey;
                }
            }
        }

        private Transition Issue(int size)
        {
            _pendingSize = Math.Max(MinInnerMtu, Math.Min(_ceiling, size));
            _pendingNonce = RandomNonce();
            _attempts = 1;
            return new Transition(new Probe(_pendingNonce, _pendingSize));
        }

        private Transition Complete(DateTimeOffset now)
        {
            _effective = Math.Max(MinInnerMtu, Math.Min(_ceiling, _effective));
            _validUntil = now + CacheTtl;
            return new Transition(CompletedMtu: _effective);
        }
    }

    private static int Normalize(int value) => Math.Clamp(value, MinInnerMtu, MaxInnerMtu);

    private static ulong RandomNonce()
    {
        Span<byte> raw = stackalloc byte[8];
        ulong nonce;
        do
        {
            RandomNumberGenerator.Fill(raw);
            nonce = BinaryPrimitives.ReadUInt64BigEndian(raw) & (ulong)long.MaxValue;
        } while (nonce == 0);
        return nonce;
    }
}
